### ISSUE TITLE
`papis doctor` and linebreaks in titles (+ other improvements)

### ISSUE THREAD

**Issue by jghauser** [papis/papis#874]
While investigating [a bug in papis.nvim](https://github.com/jghauser/papis.nvim/issues/36), I realised that the root cause might be a stray `\n` in one of the titles defined in an `info.yaml`. Are linebreaks allowed in titles? If not, it might be useful to add a check to `papis doctor` for this. The entry in question was added automatically, which means the linebreak must have snuck in via some importer/downloader.

Short list of possible random "string cleaning" improvements (also in comments):
* [ ] Remove newlines `\n` and `\n\r` from titles and abstracts, e.g. "My\nawesome paper".
* [ ] Remove repeated whitespace, e.g. "My awesome     paper".
* [ ] Remove any starting "abstract" suffix in the abstract text.
* [ ] Add some spaces in author initials, e.g. many sources give "J.R.R. Tolkien" which could be "J. R. R. Tolkien".

**Comment by alexfikl** [comment]
I'm not sure in what scenario they would be allowed, but I imagine generally it's not a good idea (it would at least mess up BibTeX).

I agree that we should have a very basic "string cleaner" to remove some stray things like that. Other things that it could do: add dots `J C -> J. C.` (mainly for authors), add spaces `J.C. -> J. C.`, remove repeated whitespace and tabs and things.

**Comment by jghauser** [comment]
Yeah that would make sense. Actually, I think any `\n` that aren't at the end of a line should be eliminated (found some more while investigating, they just happened to not be in fields that papis.nvim uses).

**Comment by avonmoll** [comment]
Slightly related question: for as long as I can remember, really long lines seem to get "hard wrapped" by papis, automatically. I can't seem to find anything about this in the docs. This almost always happens with long fields like `abstractNote`. Occasionally it happens on very long titles (which I suspect is why I was having the issues with the telescope picker in papis.nvim in the first place).

My question: is this the correct behavior for papis? It seems strange.

**Comment by alexfikl** [comment]
> Slightly related question: for as long as I can remember, really long lines seem to get "hard wrapped" by papis, automatically. I can't seem to find anything about this in the docs. This almost always happens with long fields like `abstractNote`.

@avonmoll Hm, haven't looked at this, but my best guess is that it has something to do with `yaml.dump`. At least for `pyyaml` there is a `width` argument that seems to be able to control the wrapping: https://pyyaml.org/wiki/PyYAMLDocumentation.

> Occasionally it happens on very long titles (which I suspect is why I was having the issues with the telescope picker in papis.nvim in the first place).

Does that just dump the contents of the `info.yaml` file or does it go through some YAML library? @jghauser You probably know that better?

**Comment by jghauser** [comment]
> > Occasionally it happens on very long titles (which I suspect is why I was having the issues with the telescope picker in papis.nvim in the first place).
> 
> Does that just dump the contents of the `info.yaml` file or does it go through some YAML library? @jghauser You probably know that better?

I don't think the reason for `papis.nvim` failing is due to the linebreaks inserted by Papis. I've had these as well (for instance in titles like @avonmoll), but when that happens, the `info.yaml` files still read out normally, that is, without linebreaks in the strings (I think). The problematic linebreaks are those represented as `\n` in the text (or at least that's the one that caused the error for me). So, the problematic linebreaks look like `\n` when opening the `info.yaml` file in neovim, whereas the unproblematic ones show as actual line breaks.

`papis.nvim` converts `info.yaml` files to json with `yq` and then uses neovim's inbuild json decoder to turn them into lua tables.

**Comment by jghauser** [comment]
Another improvement: Remove the word abstract (in all capitalisations) when it's the first word of the abstract (discussed in #881).
